Add release notes for 1.79.0 (#125470)
rust-lang/rust: RELEASES.md

- [Stabilize the `unnameable_types` lint for public types that can't be named.](https://github.com/rust-lang/rust/pull/120144/)
- [Enable debuginfo in macros, and stabilize `-C collapse-macro-debuginfo` and `#[collapse_debuginfo]`.](https://github.com/rust-lang/rust/pull/120845/)
- [Propagate temporary lifetime extension into `if` and `match` expressions.](https://github.com/rust-lang/rust/pull/121346/)
- [Restrict promotion of `const fn` calls.](https://github.com/rust-lang/rust/pull/121557/)
- [Split `refining_impl_trait` lint into `_reachable`, `_internal` variants.](https://github.com/rust-lang/rust/pull/121720/)
- [Stabilize associated type bounds (RFC 2289).](https://github.com/rust-lang/rust/pull/122055/)

<a id="1.79.0-Compiler"></a>

Compiler
--------
- [Define `-C strip` to only affect binaries, not artifacts like `.pdb`.](https://github.com/rust-lang/rust/pull/115120/)
- [Stabilize `-Crelro-level` for controlling runtime link hardening.](https://github.com/rust-lang/rust/pull/121694/)
- [Stabilize importing `main` from other modules or crates.](https://github.com/rust-lang/rust/pull/122060/)

> - [Stabilize checking of `cfg` names at compile-time with `--check-cfg`.](https://github.com/rust-lang/rust/pull/123501/)

Reviewer: We might want to note that Cargo is not passing this flag yet.

- [Add `aarch64-apple-visionos` and `aarch64-apple-visionos-sim` tier 3 targets.](https://github.com/rust-lang/rust/pull/121419/)
- [Add `riscv32ima-unknown-none-elf` tier 3 target.](https://github.com/rust-lang/rust/pull/122696/)

> 

Author: ```suggestion
- [Promote several Windows targets to tier 2](https://github.com/rust-lang/rust/pull/121712): `aarch64-pc-windows-gnullvm`, `i686-pc-windows-gnullvm`, and `x86_64-pc-windows-gnullvm`.

```

- [`<[T]>::split_at_mut_unchecked`](https://doc.rust-lang.org/stable/core/primitive.slice.html#method.split_at_mut_unchecked)
- [`<[u8]>::utf8_chunks`](https://doc.rust-lang.org/stable/core/primitive.slice.html#method.utf8_chunks)
- [`str::Utf8Chunks`](https://doc.rust-lang.org/stable/core/str/struct.Utf8Chunks.html)
- [`str::Utf8Chunk`](https://doc.rust-lang.org/stable/core/str/struct.Utf8Chunk.html)
- [`<*const T>::is_aligned`](https://doc.rust-lang.org/stable/core/primitive.pointer.html#method.is_aligned)
- [`<*mut T>::is_aligned`](https://doc.rust-lang.org/stable/core/primitive.pointer.html#method.is_aligned-1)
- [`NonNull::is_aligned`](https://doc.rust-lang.org/stable/core/ptr/struct.NonNull.html#method.is_aligned)
- [`<*const [T]>::len`](https://doc.rust-lang.org/stable/core/primitive.pointer.html#method.len)
- [`<*mut [T]>::len`](https://doc.rust-lang.org/stable/core/primitive.pointer.html#method.len-1)
- [`<*const [T]>::is_empty`](https://doc.rust-lang.org/stable/core/primitive.pointer.html#method.is_empty)
- [`<*mut [T]>::is_empty`](https://doc.rust-lang.org/stable/core/primitive.pointer.html#method.is_empty-1)

> - [`NonNull::is_empty`](https://doc.rust-lang.org/stable/core/ptr/struct.NonNull.html#method.is_empty)

Reviewer 2: should be `NonNull<[T]>::is_empty`

Author: ```suggestion
- [`NonNull::<[T]>::is_empty`](https://doc.rust-lang.org/stable/core/ptr/struct.NonNull.html#method.is_empty)
```

Version 1.79.0 (2024-06-13)
==========================

> <!-- FIXME: unsorted T-types relnotes -->
> - [check `FnDef` return type for WF](https://github.com/rust-lang/rust/pull/115538/)
> - [rework opaque type region inference](https://github.com/rust-lang/rust/pull/116891/)
> - [instantiate higher ranked goals outside of candidate selection](https://github.com/rust-lang/rust/pull/119820/)
> - [Pass list of defineable opaque types into canonical queries](https://github.com/rust-lang/rust/pull/122077/)
> - [Eagerly instantiate closure/coroutine-like bounds with placeholders to deal with binders correctly](https://github.com/rust-lang/rust/pull/122267/)
> - [Use fulfillment in method probe, not evaluation](https://github.com/rust-lang/rust/pull/122317/)
> - [Make inductive cycles always ambiguous](https://github.com/rust-lang/rust/pull/122791/)
> - [Enforce closure args + return type are WF](https://github.com/rust-lang/rust/pull/123531/)
> - [Fix trait solver overflow with `non_local_definitions` lint](https://github.com/rust-lang/rust/pull/123594/)
> - [More DefineOpaqueTypes::Yes](https://github.com/rust-lang/rust/pull/123794/)
> 

Reviewer 3: ```suggestion
```

<a id="1.79.0-Language"></a>

Language
--------
- [Stabilize inline `const {}` expressions.](https://github.com/rust-lang/rust/pull/104087/)
- [Prevent opaque types being instantiated twice with different regions within the same function.](https://github.com/rust-lang/rust/pull/116935/)
- [Stabilize WebAssembly target features that are in phase 4 and 5.](https://github.com/rust-lang/rust/pull/117457/)
- [Add the `redundant_lifetimes` lint to detect lifetimes which are semantically redundant.](https://github.com/rust-lang/rust/pull/118391/)
- [Stabilize the `unnameable_types` lint for public types that can't be named.](https://github.com/rust-lang/rust/pull/120144/)
- [Enable debuginfo in macros, and stabilize `-C collapse-macro-debuginfo` and `#[collapse_debuginfo]`.](https://github.com/rust-lang/rust/pull/120845/)
- [Propagate temporary lifetime extension into `if` and `match` expressions.](https://github.com/rust-lang/rust/pull/121346/)
- [Restrict promotion of `const fn` calls.](https://github.com/rust-lang/rust/pull/121557/)

> - [Split `refining_impl_trait` lint into `_reachable`, `_internal` variants.](https://github.com/rust-lang/rust/pull/121720/)

Reviewer 4: Specifically, the change as compared with stable Rust is that we're now linting refining impls of traits that are not reachable from outside the crate when we were not doing that before.

Perhaps most explicitly:

```suggestion
- [Rename the existing `refining_impl_trait` lint to `refining_impl_trait_reachable`; add `refining_impl_trait_internal` lint to warn against impls that refine a trait not reachable from outside the crate; place these both in a `refining_impl_trait` lint group.](https://github.com/rust-lang/rust/pull/121720/)
```

Or maybe more concisely:

```suggestion
- [Warn against refining impls of crate-private traits with `refining_impl_trait` lint.](https://github.com/rust-lang/rust/pull/121720/)
```

- [Stabilize associated type bounds (RFC 2289).](https://github.com/rust-lang/rust/pull/122055/)

> 

Reviewer 3: ```suggestion
- [Check return types of function types for well-formedness](https://github.com/rust-lang/rust/pull/115538)
- [Rework `impl Trait` lifetime inference](https://github.com/rust-lang/rust/pull/116891/)
- [Change inductive trait solver cycles to be ambiguous](https://github.com/rust-lang/rust/pull/122791)

```

Reviewer 5: The first of those at least seems like it should be in compat notes? Not sure though. Maybe both?

> 

Reviewer 4: ```suggestion
- [Stabilize importing `main` from other modules or crates.](https://github.com/rust-lang/rust/pull/122060/)

```

This was down in the compiler section but is really a lang feature.

<a id="1.79.0-Compiler"></a>

Compiler
--------
- [Define `-C strip` to only affect binaries, not artifacts like `.pdb`.](https://github.com/rust-lang/rust/pull/115120/)
- [Stabilize `-Crelro-level` for controlling runtime link hardening.](https://github.com/rust-lang/rust/pull/121694/)

> - [Stabilize importing `main` from other modules or crates.](https://github.com/rust-lang/rust/pull/122060/)

Reviewer 4: ```suggestion
```

This is really a lang feature.
